### PR TITLE
[FSTORE-1186][APPEND] Polars Integration into Hopsworks - Writing and Reading DataFrames

### PR DESCRIPTION
**Pull request by manu-sj**
This PR fixes two issue present in the integration of Polars to Hopsworks

1. **Helper Columns** - When reading helper columns as a polars dataframe incorrect columns with empty data was obatined.

- _Root Cause_: Issue caused due to incorrect schema being used to create the polars dataframe. Polars dataframe was created using schema extracted from feature vectors.
- _Fix done_ : Corrected the schema used by extracting the inference helper columns.

3. **Training Data** - When creating training data from a feature group with very little data, an exception is encountered sporadically.

- _Root Cause_ : Issue caused due to the very little data being available in the feature group. When very little training data is present in the feature group there is a possibility that one of the training data split (train, test or validation) can be empty. Spark writes an empty file in this case. When reading from polars empty file were not read into a list since it causes an exception to be throw while merging into a single dataframe due to schema mismatch. Hence a empty list would be obtained if the entire split is empty which causes an exception when trying to merge with split.
- _Fix done_ : Check added to see if df list is empty and return an empty polars dataframe with column names in that usecase.

**How Has This Been Tested?**

- [x] Unit Tests
- [ ] Integration Tests
- [x] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: python/hsfs/engine/python.py [hunk 1]
```diff
@@ -321,7 +328,9 @@ def _read_hopsfs(
         try:
             from pydoop import hdfs
         except ModuleNotFoundError:
-            return self._read_hopsfs_remote(location, data_format, read_options or {})
+            return self._read_hopsfs_remote(
+                location, data_format, read_options, dataframe_type
```
Comment (vatj) [inline python/hsfs/engine/python.py:332]: Why did you remove the `or {}` part?

File: python/hsfs/engine/python.py [hunk 2]
```diff
@@ -364,13 +371,20 @@ def _read_hopsfs_remote(self, location, data_format, read_options=None):
                     ):
                         arrow_flight_config = read_options.get("arrow_flight_config")
                         df = arrow_flight_client.get_instance().read_path(
-                            inode.path, arrow_flight_config
+                            inode.path,
+                            arrow_flight_config,
+                            dataframe_type=dataframe_type,
```
Comment (vatj) [inline python/hsfs/engine/python.py:376]: Why do you need to pass dataframe type here and use it again below?

File: python/hsfs/engine/spark.py [hunk 3]
```diff
@@ -532,8 +532,12 @@ def get_training_data(
 
     def split_labels(self, df, labels, dataframe_type):
         if labels:
-            labels_df = df.select(*labels)
-            df_new = df.drop(*labels)
+            if isinstance(df, pd.DataFrame):
```
Comment (vatj) [inline python/hsfs/engine/spark.py:535]: Why is this needed? Was it not working before? Or do you need this for polars instead of pandas?

File: python/tests/engine/test_python.py [hunk 4]
```diff
@@ -249,6 +249,68 @@ def test_read_hopsfs_connector(self, mocker):
...
 
+    def test_read_hopsfs_connector_empty_dataframe(self, mocker):
+        # Arrange
+
+        # Setting list of empty dataframes as return value from _read_hopsfs
+        mock_python_engine_read_hopsfs = mocker.patch(
+            "hsfs.engine.python.Engine._read_hopsfs",
+            return_value=[pd.DataFrame(), pd.DataFrame()],
+        )
+        mock_python_engine_read_s3 = mocker.patch("hsfs.engine.python.Engine._read_s3")
```
Comment (vatj) [inline python/tests/engine/test_python.py:260]: Why do we need to patch S3 here?

File: python/tests/engine/test_python.py [hunk 5]
```diff
@@ -249,6 +249,68 @@ def test_read_hopsfs_connector(self, mocker):
...
+            storage_connector=connector,
+            data_format="csv",
+            read_options=None,
+            location=None,
+            dataframe_type="default",
+        )
+
+        # Assert
+        assert mock_python_engine_read_hopsfs.call_count == 1
+        assert mock_python_engine_read_s3.call_count == 0
```
Comment (vatj) [inline python/tests/engine/test_python.py:279]: Why do we need to check that we don't read from S3?

File: python/tests/engine/test_python.py [hunk 6]
```diff
@@ -249,6 +249,68 @@ def test_read_hopsfs_connector(self, mocker):
...
+
+    def test_read_hopsfs_connector_empty_dataframe_polars(self, mocker):
+        # Arrange
+
+        # Setting empty list as return value from _read_hopsfs
+        mock_python_engine_read_hopsfs = mocker.patch(
+            "hsfs.engine.python.Engine._read_hopsfs",
+            return_value=[pl.DataFrame(), pl.DataFrame()],
+        )
+        mock_python_engine_read_s3 = mocker.patch("hsfs.engine.python.Engine._read_s3")
```
Comment (vatj) [inline python/tests/engine/test_python.py:291]: Same?

File: python/hsfs/engine/python.py [hunk 7]
```diff
@@ -274,9 +274,16 @@ def read(
...
+            # Below check performed since some files materialized when creating training data are empty
+            # If empty dataframe is in df_list then polars cannot concatenate df_list due to schema mismatch
+            # However if the entire split contains only empty files which can occur when the data size is very small then one of the empty dataframe is return so that the column names can be accessed.
+            non_empty_df_list = [df for df in df_list if not df.is_empty()]
+            if non_empty_df_list:
+                return self._return_dataframe_type(
+                    pl.concat(non_empty_df_list), dataframe_type=dataframe_type
+                )
+            else:
+                return df_list[0]
```
[the remark this answers is not in the document]
Comment (manu-sj) [reply python/hsfs/engine/python.py:286]: Yes that is the problem. Since there is no data polars cannot infer the type. Pandas use the "object" type in these usecases.

File: python/hsfs/engine/python.py (hunk 1 above)
Comment (manu-sj) [reply python/hsfs/engine/python.py:332]: Merging mistake sorry corrected.

File: python/hsfs/engine/python.py (hunk 2 above)
Comment (manu-sj) [reply python/hsfs/engine/python.py:376]: Arrow flight returns the dataframe as pandas by default. I had modifed `_get_dataset` function in _arrow_flight_client.py_ to read as polars dataframe when the dataframe type is polars. So this passing of `dataframe_type` is used to acheive this functionality.

File: python/tests/engine/test_python.py (hunk 4 above)
Comment (manu-sj) [reply python/tests/engine/test_python.py:260]: Not needed removed.

File: python/tests/engine/test_python.py (hunk 5 above)
Comment (manu-sj) [reply python/tests/engine/test_python.py:279]: Not needed removed.

File: python/tests/engine/test_python.py (hunk 6 above)
Comment (manu-sj) [reply python/tests/engine/test_python.py:291]: Not needed removed.

File: python/hsfs/core/vector_server.py [hunk 8]
```diff
@@ -373,7 +378,7 @@ def get_inference_helper(self, entry, return_type):
         elif return_type.lower() == "polars":
             # Polar considers one dimensional list as a single columns so passed vector as 2d list
             polars_df = pl.DataFrame(
-                [serving_vector], schema=self._feature_vector_col_name, orient="row"
+                [serving_vector], schema=self._inference_helper_col_name, orient="row"
```
[the remark this answers is not in the document]
Comment (manu-sj) [reply python/hsfs/core/vector_server.py]: Changed it to use serving_vector.keys() for `get_inference_helper`.

For `get_inference_helpers` since `batch_results` is a list I used `feature_view_object.inference_helper_columns` which was a pre-existing variable in feature view. I am wondering if I should use it for  `get_inference_helper` too?

File: python/hsfs/engine/spark.py (hunk 3 above)
Comment (vatj) [reply python/hsfs/engine/spark.py:535]: "This behaviour was changed during the implementation of Polars. And dataframe were directly read as Pandas/Polars and Spark dataframe based on the dataframe_type."
It seems like it could break a lot of things. You mean that now you do things that were being done with spark datafra me directly using polars or pandas dataframe?

File: python/hsfs/core/vector_server.py [hunk 9]
```diff
@@ -373,7 +373,7 @@ def get_inference_helper(self, entry, return_type):
         elif return_type.lower() == "polars":
             # Polar considers one dimensional list as a single columns so passed vector as 2d list
             polars_df = pl.DataFrame(
-                [serving_vector], schema=self._feature_vector_col_name, orient="row"
+                [serving_vector], schema=serving_vector.keys(), orient="row"
```
Comment (vatj) [inline python/hsfs/core/vector_server.py:376]: Actually can serving vector be None?

**Review by vatj — APPROVED** [review]
